src/StackExchange.Redis/ChannelMessageQueue.cs

```diff
@@ -162,17 +175,29 @@ private async Task OnMessageSyncImpl()
             while (!Completion.IsCompleted)
             {
                 ChannelMessage next;
-                try { if (!TryRead(out next)) next = await ReadAsync().ConfigureAwait(false); }
-                catch (ChannelClosedException) { break; } // expected
-                catch (Exception ex)
+
+                // Keep trying to read values
+                while (!TryRead(out next))
                 {
-                    _parent.multiplexer?.OnInternalError(ex);
-                    break;
+                    // If we fail, wait for an item to appear
+                    if (!await _queue.WaitToReadAsync())
+                    {
+                        // Channel is closed
+                        break;
+                    }
+
+                    // There should be an item available now, but another reader might grab it,
+                    // so we keep TryReading in the loop.
                 }
 
                 try { handler(next); }
                 catch { } // matches MessageCompletable
             }
+
+            if (Completion.IsFaulted)
+            {
+                _onInternalError(Completion.Exception.InnerException);
```

Comment:
This is technically new behavior, it will trigger this callback if the channel is closed with an `Exception`, but we don't ever do that in the normal flow. A custom implementor that uses their own Channel in `ChannelMessageQueue` might though...